apps/google-analytics-4/lambda/src/services/googleApiUtils.ts

```diff
@@ -0,0 +1,147 @@
+import { ERROR_TYPE_MAP } from '../errors/apiError';
+import { Status } from 'google-gax';
+import { HttpCodeToRpcCodeMap } from 'google-gax/build/src/status';
+
+interface GoogleApiErrorType {
+  details: string;
+  status: number;
+  errorType: string;
+}
+
+export class GoogleApiError extends Error {
+  details: string;
+  status: number;
+  errorType: string;
+
+  constructor(
+    message: ConstructorParameters<typeof Error>[0],
+    options: ConstructorParameters<typeof Error>[1],
+    errorParams: GoogleApiErrorType
+  ) {
+    super(message, options);
+    this.errorType = errorParams.errorType;
+    this.details = errorParams.details;
+    this.status = errorParams.status;
+  }
+}
+
+const clientErrorStatuses = [
+  Status.INVALID_ARGUMENT,
+  Status.UNAUTHENTICATED,
+  Status.PERMISSION_DENIED,
+  Status.NOT_FOUND,
+  Status.ABORTED,
+  Status.OUT_OF_RANGE,
+  Status.RESOURCE_EXHAUSTED,
+  Status.CANCELLED,
+  Status.FAILED_PRECONDITION,
+];
+
+const httpStatusFromGoogleRpcStatus = (status: Status, fallbackStatus = 500): number => {
+  for (const [httpStatus, googleStatus] of HttpCodeToRpcCodeMap) {
+    if (googleStatus === status) {
+      return httpStatus;
+    }
+  }
+  return fallbackStatus;
+};
+
+export function createThrowableErrorFromCommonErrors(e: any) {
+  if (e.code === Status.INVALID_ARGUMENT) {
+    return new GoogleApiError(
+      e.message,
+      { cause: e },
+      {
+        details: e.message,
+        status: httpStatusFromGoogleRpcStatus(e.code),
+        errorType: ERROR_TYPE_MAP.noAccountsOrPropertiesFound,
+      }
+    );
+  } else if (e.code === Status.UNAUTHENTICATED) {
+    return new GoogleApiError(
+      e.message,
+      { cause: e },
+      {
+        details: e.message,
+        status: httpStatusFromGoogleRpcStatus(e.code),
+        errorType: ERROR_TYPE_MAP.invalidServiceAccount,
+      }
+    );
+  } else {
+    return new GoogleApiError(
+      e.message,
+      { cause: e },
+      {
+        details: e.message,
+        status: httpStatusFromGoogleRpcStatus(e.code),
+        errorType: ERROR_TYPE_MAP.unknown,
+      }
+    );
+  }
+}
+
+export function handleGoogleAdminApiError(e: any): never {
+  if (e.code === Status.PERMISSION_DENIED) {
+    throw new GoogleApiError(
+      e.message,
+      { cause: e },
+      {
+        details: e.message,
+        status: httpStatusFromGoogleRpcStatus(e.code),
+        errorType: ERROR_TYPE_MAP.disabledAdminApi,
+      }
+    );
+  } else {
+    const throwableError = createThrowableErrorFromCommonErrors(e);
+    throw throwableError;
+  }
+}
+
+export function handleGoogleDataApiError(e: any): never {
+  if (e.code === Status.PERMISSION_DENIED) {
```

Comment:
Hmm. I'm concerned this is a false assumption. We should at minimum verify this ourselves!

My hypothesis would be there are at minimum two cases where you would see this error:

* The project does not have the google data api enabled
* The project DOES have the api enabled but the requesting user does not have permission to review the data (i.e. run the report)